Provides Convenient Lambda Overload Visitor Interface, resolves #113.

daniel-j-h opened the pull request:
See https://github.com/mapbox/variant/issues/113 for details.

This provides a `make_visitor` function taking lambdas and creating a
visitor functor out of it.

There are three points to conder here:

1/ Eventually we want to provide a convenient member function
so that we can plug in lambda overloads directly as in:

```c++
variant.match([] (Response) { return "ok"; },
              [] (Error)    { return "error"; });
```

at the moment we still have to:

```c++
auto visitor = make_visitor([] (Response) { return "ok"; },
                            [] (Error)    { return "error"; });
apply_visitor(visitor, variant);
```

2/ At the moment we copy the lambdas. This is fine and follows the
stdlib but we could take them as forwarding reference instead.

3/ Maybe we should consider edge cases, such as member function pointers
and provide specializations accordingly. The primary use-case is lambdas,
though.

tomilov commented:
Do you really need `visitor<Fns...>::type` instead of `visitor<Fns...>` everywhere?

daniel-j-h commented:
Hm you are right for our use-case this _should_ work, too. Changed in https://github.com/mapbox/variant/pull/124/commits/2275a61974aaf117fa8d08c08d640ffd05935db8.

I also added a test for SFINAE-based introspection for lambdas in https://github.com/mapbox/variant/pull/124/commits/4d462f27b2d852f66a3769e5983691c0f9233c9e.

Works quite nicely for generic lambda overload sets.

artemp commented:
@daniel-j-h - looking good, cheers! 

@joto - what's the motivation for keeping unit tests in `test/t` ? Should we port all test there so they can be run via `make test` ?

joto commented:
@artemp Having the actual test code in `test/t` instead of just `test` is just a way of organizing the repository. Usually there is other "stuff" hanging around in a `test` directory, such as a README, data files, Makefiles, etc. So I put the actual code in `t`. Just like putting the code under `src` instead of having it in the root directory of a repository. Some tests don't work as unit tests in the same way, in this case the stuff in `compilation_failure`, because it doesn't compile. There is a different script to run those (`scripts/run_compilation_failure_tests.sh`). The source code that's not in `t` or `compilation_failure` probably needs to be turned into proper unit test.

artemp commented:
@joto - ok, great, makes sense.

tomilov commented:
Member typedef `type` makes sense for unary and (slightly less) for binary case. For unary case `make_visitor` returns its sole argument. For binary there is only single layer of inheritance.
